npeps/npep-133-fqdn-egress-selector.md

```diff
@@ -0,0 +1,442 @@
+line.
+
+* As a cluster admin, I want to block all cluster egress traffic by default, and
+  require namespace admins to create NetworkPolicies explicitly allowing egress
+  to the domains they need to talk to.
+
+  The Cluster admin would use a `BaselineAdminNetworkPolicy` object to switch
+  the default disposition of the cluster. Namespace admins would then use a FQDN
+  selector in the Kubernetes `NetworkPolicy` objects to allow `my-service.com`.
+  
+## API
+
+This NPEP proposes adding a new type of `AdminNetworkPolicyEgressPeer` called
+`FQDNPeerSelector` which allows specifying domain names.
+
+```golang
+type AdminNetworkPolicyEgressPeer struct {
+    <snipped>
+    // FQDN provides a way to specify domain names as peers.
+    //
+    // FQDN is only supported for ALLOW rules. In order to control access, FQDN
+    // ALLOW rules should be used with a lower priority egress deny -- this
+    // allows the admin to maintain an explicit "allowlist" of reachable
+   // domains.
+    //
+    // Support: Extended
+    //
+    // +optional
+    FQDN *FQDNPeerSelector `json:"fqdn,omitempty"`
+}
+
+type FQDNPeerSelector struct {
+    // Name specifies the exact Fully-Qualified Domain Name to select (the
+    // trailing '.' is optional)
+    //
+    // This field acts as an exact match. It doesn't match any other
+    // domains of that level or even any subdomains.
+    // Examples:
+    //   - `kubernetes.io` matches only `kubernetes.io`.
+    //     It does not match "www.kubernetes.io", "blog.kubernetes.io",
+    //     "my-kubernetes.io", or "wikipedia.org".
+    //   - `blog.kubernetes.io` matches only "blog.kubernetes.io".
+    //     It does not match "www.kubernetes.io" or "kubernetes.io".
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +kubebuilder:validation:Pattern=`^(([a-zA-z0-9]|[a-zA-Z0-9][-a-zA-Z0-9_]*[a-zA-Z0-9])\.)*([a-zA-z0-9]|[a-zA-Z0-9][-a-zA-Z0-9_]*[a-zA-Z0-9])*\.?$`
+    Name string `json:"names,omitempty"`
+
+    //  Pattern allows matching FQDNs using wildcard specifiers.
+    //
+    // "*" matches 0 or more DNS valid characters (except for "."), and may
+    // occur anywhere in the pattern.
+    //  Examples:
+    //    - `*.kubernetes.io` matches subdomains of kubernetes.io at that level.
+    //      "www.kubernetes.io" and "blog.kubernetes.io" match, however
+    //      "kubernetes.io", "latest.blog.kubernetes.io", and "wikipedia.org" do
+    //      not.
+    //    - `*.*.kubernetes.io` matches subdomains of kubernetes.io at that
+    //      level.
+    //      "sub.subdomain.kubernetes.io" matches, however "kubernetes.io" and
+    //      "blog.kubernetes.io" do not.
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +kubebuilder:validation:Pattern=`^((\*|[a-zA-z0-9]|[a-zA-Z0-9][-a-zA-Z0-9_]*[a-zA-Z0-9])\.)*(\*|[a-zA-z0-9]|[a-zA-Z0-9][-a-zA-Z0-9_]*[a-zA-Z0-9])*\.?$`
+    Pattern string `json:"pattern,omitempty"`
```

Comment:
Calico (Enterprise) has DNS policy.  Some thoughts from our impl:

* We only have one field for domain vs pattern so that'd be my preference.
* We only support a single wildcard in the domain but it may match more than one dot-separated part.  
* I think we disallow a redundant trailing `.`.  I think that trailing dot has a subtle meaning in some contexts but since we're not honouring that subtle meaning here I say avoid redundancy.

Our wildcarding works like this:

- *.google.com matches www.google.com and www.ipv6.google.com, but not google.com
- www.*.com matches www.sun.com and www.apple.com, but not www.com
- www.google.* matches www.google.net, www.google.io, and so on
- update.*.mycompany.com matches update.tools.mycompany.com, update.secure.suite.mycompany.com, and so on

We don't allow 

- Multiple wildcards in the same domain, for example: *.*.mycompany.com
- Asterisks that are not the entire component, for example: www.g*.com
- More general wildcards, such as regular expressions

I think those limitations were mainly UI driven but the limit of * only matching a whole segment might allow for more efficient indexing (as opposed to needing to do a regex).

Reply:
Sounds good to me. For the most part I believe this is compatible with what we're proposing.